File: scripts/generate.mjs
import {
  serviceContextFromDefaults,
  storageContextFromDefaults,
  SimpleDirectoryReader,
  VectorStoreIndex,
} from "llamaindex";

import {
  DATASOURCES_CACHE_DIR,
  DATASOURCES_DIR,
  DATASOURCES_CHUNK_SIZE,
  DATASOURCES_CHUNK_OVERLAP,
} from "./constants.mjs";
import { exit } from "process";

async function getRuntime(func) {
  const start = Date.now();
  await func();
  const end = Date.now();
  return end - start;
}

async function generateDatasource(serviceContext, datasource) {
  console.log(`Generating storage context for datasource '${datasource}'...`);
  // Split documents, create embeddings and store them in the storage context
  const ms = await getRuntime(async () => {
    const storageContext = await storageContextFromDefaults({
      persistDir: `${DATASOURCES_CACHE_DIR}/${datasource}`,
    });
    const documents = await new SimpleDirectoryReader().loadData({
      directoryPath: `${DATASOURCES_DIR}/${datasource}`,
    });
    await VectorStoreIndex.fromDocuments(documents, {
      storageContext,
      serviceContext,
    });
  });
  console.log(
    `Storage context for datasource '${datasource}' successfully generated in ${
      ms / 1000
    }s.`,
  );
}

const datasource = process.argv[2];

if (!datasource) {
  console.log("Error: You must provide a datasource as the parameter.");
  console.log("Usage: pnpm run generate <datasource>");
  exit(1);
}

(async () => {
  const serviceContext = serviceContextFromDefaults({
    chunkSize: DATASOURCES_CHUNK_SIZE,
    chunkOverlap: DATASOURCES_CHUNK_OVERLAP,
  });

  await generateDatasource(serviceContext, datasource);
  console.log("Finished generating datasource.");
})();

File: backend/generate.py
import sys
from llama_index import (
    SimpleDirectoryReader,
    VectorStoreIndex,
    ServiceContext,
)
from app.constants import (
    DATASOURCES_CACHE_DIR,
    DATASOURCES_DIR,
    DATASOURCES_CHUNK_SIZE,
    DATASOURCES_CHUNK_OVERLAP,
)


def generateDatasource(service_context: ServiceContext, datasource: str):
    ds_data_dir = DATASOURCES_DIR + "/" + datasource
    ds_storage_dir = DATASOURCES_CACHE_DIR + "/" + datasource
    print(f"Generating storage context for datasource '{datasource}'...")
    # load the documents and create the index
    documents = SimpleDirectoryReader(ds_data_dir).load_data()
    index = VectorStoreIndex.from_documents(
        documents, service_context=service_context, show_progress=True
    )
    # store it for later
    index.storage_context.persist(ds_storage_dir)
    print(f"Finished creating new index. Stored in {ds_storage_dir}")


def main():
    if len(sys.argv) < 2:
        print("Please provide a datasource parameter.")
        return

    datasource = sys.argv[1]
    service_context = ServiceContext.from_defaults(
        chunk_size=DATASOURCES_CHUNK_SIZE, chunk_overlap=DATASOURCES_CHUNK_OVERLAP
    )
    generateDatasource(service_context, datasource)
    print("Finished generating datasource.")


if __name__ == "__main__":
    main()
